quantum/action_util.c

```diff
@@ -295,6 +298,53 @@ void send_keyboard_report(void) {
 #endif
 }
 
+#ifdef NKRO_ENABLE
+void send_nkro_report(void) {
+    nkro_report->mods = real_mods | weak_mods;
+
+#    ifndef NO_ACTION_ONESHOT
+    if (oneshot_mods) {
+#        if (defined(ONESHOT_TIMEOUT) && (ONESHOT_TIMEOUT > 0))
+        if (has_oneshot_mods_timed_out()) {
+            dprintf("Oneshot: timeout\n");
+            clear_oneshot_mods();
+        }
+#        endif
+        nkro_report->mods |= oneshot_mods;
+        if (has_anykey()) {
+            clear_oneshot_mods();
+        }
+    }
+#    endif
+
+#    ifdef KEY_OVERRIDE_ENABLE
+    // These need to be last to be able to properly control key overrides
+    nkro_report->mods &= ~suppressed_mods;
+    nkro_report->mods |= weak_override_mods;
+#    endif
+
+    static report_nkro_t last_report;
+
+    /* Only send the report if there are changes to propagate to the host. */
+    if (memcmp(nkro_report, &last_report, sizeof(report_nkro_t)) != 0) {
+        memcpy(&last_report, nkro_report, sizeof(report_nkro_t));
+        host_nkro_send(nkro_report);
+    }
+}
+#endif
+
+void send_keyboard_report(void) {
+#ifdef NKRO_ENABLE
+    if (keymap_config.nkro) {
```

Comment:
```suggestion
    if (keyboard_protocol && keymap_config.nkro) {
```
The condition used here must exactly match the one in `tmk_core/protocol/report.c`, otherwise a wrong kind of report will be sent.  The problematic case here is when the keyboard has the NKRO mode toggled on in the EEPROM, but the host switches the keyboard into the boot protocol mode (e.g., when you try to enter BIOS and change some settings, or even interact with the bootloader before the full OS with proper HID drivers gets loaded).

(Looks like the atsam code does not handle that part at all — maybe it works in the boot protocol mode only when NKRO is not enabled.)

Maybe the proper way is to refactor this condition into some function, like `bool is_nkro_active(void)`.  Then this function could also be used to handle other cases when the 6KRO mode needs to be forced (e.g., if the NKRO mode is desired for the wired connection, but the wireless connection supports only 6KRO).  Some proper functions for switching between modes would probably need to be added too (the current code mostly ignores that possibility, but if the boot protocol gets enabled or disabled while some keys are pressed, the reports might contain garbage; just clearing the report on switch would probably be good enough).